### PR TITLE
Workflow metrics

### PR DESCRIPTION
**Pull request by hoodakaushal**
Added more detailed metrics for workflows - a rate measure for each individual type of workflow being triggered.

### CODE REVIEW

File: runtime/src/main/java/com/flipkart/flux/guice/module/ContainerModule.java [hunk 1]
```diff
@@ -65,7 +67,7 @@
 	 */
 	@Override
 	protected void configure() {
-
+		bind(MetricsClient.class).to(MetricsClientImpl.class);
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/guice/module/ContainerModule.java]: Make it singleton
bind(MetricsClient.class).to(MetricsClientImpl.class).in(Singleton.class);

File: runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package com.flipkart.flux.metrics.impl;
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java]: Add Copyright, refer to any file for the copyright text. You can configure Intellij to add copyright on creation.
https://www.jetbrains.com/help/idea/2016.3/generating-and-updating-copyright-notice.html

File: runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.codahale.metrics.MetricRegistry;
+import com.flipkart.flux.metrics.interfaces.MetricsClient;
+import com.google.inject.Inject;
+
+/**
+ * Created by kaushal.hooda on 25/01/17.
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java]: Add documentation

File: runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Inject
+    MetricsClientImpl(MetricRegistry metricRegistry){
+        this.metricRegistry = metricRegistry;
+    }
+
+
+
+    @Override
+    public void incrCount(String key) {
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java]: Remove unused methods

File: runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        metricRegistry.counter(key).inc();
+    }
+
+    @Override
+    public void decrCount(String key) {
+        metricRegistry.counter(key).dec();
+    }
+
+    @Override
+    public void markEvent(String key) {
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/metrics/impl/MetricsClientImpl.java]: markMeter?

File: runtime/src/main/java/com/flipkart/flux/metrics/interfaces/MetricsClient.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.flipkart.flux.metrics.interfaces;
+
+/**
+ * Created by kaushal.hooda on 25/01/17.
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/metrics/interfaces/MetricsClient.java]: Add documentation

File: runtime/src/main/java/com/flipkart/flux/resource/StateMachineResource.java [hunk 7]
```diff
@@ -123,6 +128,7 @@ public Response createStateMachine(StateMachineDefinition stateMachineDefinition
             return Response.status(Response.Status.CONFLICT.getStatusCode()).entity(ex.getCause() != null ? ex.getCause().getMessage() : null).build();
         }
 
+        metricsClient.markEvent(stateMachine.getName());
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/resource/StateMachineResource.java]: move this up into try block

File: runtime/src/main/java/com/flipkart/flux/resource/StateMachineResource.java [hunk 8]
```diff
@@ -13,6 +13,7 @@
 
 package com.flipkart.flux.resource;
 
+import com.codahale.metrics.Metric;
```
Comment (shyamakirala) [inline runtime/src/main/java/com/flipkart/flux/resource/StateMachineResource.java]: unused import

**Comment by shyamakirala** [comment]
Also move the below provider method from RedriverModule to ContainerModule, as it seems apt to keep it there. 

public MetricRegistry metricRegistry()

**Comment by hoodakaushal** [comment]
Made the suggested changes.

**Pull request merged**